runner/src/recipe-manager.ts

```diff
@@ -0,0 +1,327 @@
+    if (!recipeMeta.src) {
+      console.error(
+        "registerRecipe: no reason to save recipe, missing src",
+        recipeId,
+      );
+      return false;
+    }
+
+    // FIXME(ja): should we update the recipeMeta if it already exists? when does this happen?
+    if (recipeMetaMap.has(recipe as Recipe)) {
+      return true;
+    }
+
+    const recipeMetaCell = await this.getRecipeMetaCell({ recipeId, space });
+    recipeMetaCell.set(recipeMeta);
+    recipeMetaMap.set(recipe as Recipe, recipeMetaCell);
+    await storage.syncCell(recipeMetaCell);
+    await storage.synced();
+
+    recipeIdMap.set(recipeId, recipe as Recipe);
+    recipeMetaMap.set(recipe as Recipe, recipeMetaCell);
+
+    // FIXME(ja): in a week we should remove auto-publishing to blobby
+    // if this patch doesn't need to be reverted
+    await this.publishToBlobby(recipeId);
+
+    return true;
+  }
+
+  // returns a recipe already loaded
+  recipeById(recipeId: string): Recipe | undefined {
+    return recipeIdMap.get(recipeId);
+  }
+
+  // we need to ensure we only compile once otherwise we get ~12 +/- 4
+  // compiles of each recipe
+  private async compileRecipeOnce(
+    recipeId: string,
+    space: string,
+  ): Promise<Recipe> {
+    const metaCell = await this.getRecipeMetaCell({ recipeId, space });
+    let recipeMeta = metaCell.get();
+
+    // 1. Fallback to Blobby if cell missing or stale
+    if (recipeMeta.id !== recipeId) {
+      const imported = await this.importFromBlobby({ recipeId });
+      recipeMeta = imported.recipeMeta;
+      metaCell.set(recipeMeta);
+      await storage.syncCell(metaCell);
+      await storage.synced();
+      recipeIdMap.set(recipeId, imported.recipe);
+      recipeMetaMap.set(imported.recipe, metaCell);
+      return imported.recipe;
+    }
+
+    // 2. Compile from stored source
+    if (!recipeMeta.src) {
+      throw new Error(`Recipe ${recipeId} has no stored source`);
+    }
+    const recipe = await runtime.compile(recipeMeta.src);
+
+    metaCell.set(recipeMeta);
+    await storage.syncCell(metaCell);
+    await storage.synced();
+    recipeIdMap.set(recipeId, recipe);
+    recipeMetaMap.set(recipe, metaCell);
+    return recipe;
+  }
+
+  async loadRecipe(
+    { space, recipeId }: { space: string; recipeId: string },
+  ): Promise<Recipe> {
+    // quick paths
+    const existingRecipe = recipeIdMap.get(recipeId);
+    if (existingRecipe) {
+      return existingRecipe;
+    }
+
+    if (inProgressCompilations.has(recipeId)) {
+      return inProgressCompilations.get(recipeId)!;
+    }
+
+    // single-flight compilation
+    const compilationPromise = this.compileRecipeOnce(recipeId, space)
+      .finally(() => inProgressCompilations.delete(recipeId)); // tidy up
+
+    inProgressCompilations.set(recipeId, compilationPromise);
+    return await compilationPromise;
+  }
+
+  /**
+   * Load a recipe from Blobby, returning the recipe and recipeMeta
+   */
+  // FIXME(ja): move this back to blobby!
```

Comment:
Is this fixed?

Reply:
oh, I was thinking the logical fetch get/put against blobby maybe should be back in blobby-storage.ts - since they are mostly self-contained...

or at least these parts:

get

```
    const response = await fetch(`${getBlobbyServerUrl()}/spell-${recipeId}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch recipe ${recipeId} from blobby`);
    }

    const recipeJson = await response.json() as {
      src: string;
      spec?: string;
      parents?: string[];
    };
```

and all/the majority of publishToBlobby 

---

but I don't think we need to decide this for this PR to land